io/src/main/scala/fs2/io/ToInputStream.scala

```diff
@@ -0,0 +1,259 @@
+  */
+object ToInputStream {
+
+
+
+  def apply[F[_]](implicit F:Async[F]):Pipe[F,Byte,InputStream] = {
+    /*
+     * Implementation note:
+     *
+     * We run this through 3 synchronous primitives
+     *
+     * - Synchronous Queue -  used to signal next available chunk, or when the upstream is done/failed
+     * - UpStream signal -    used to monitor state of upstream, primarily to indicate to `close`
+     *                        that upstream has finished and is safe time to terminate
+     * - DownStream signal -  keeps any remainders from last `read` and signals
+     *                        that downstream has been terminated that in turn kills upstream
+     *
+     */
+    (source:Stream[F,Byte]) =>
+      Stream.eval(async.synchronousQueue[F,Either[Option[Throwable],Bytes]]).flatMap { queue =>
+        Stream.eval(async.signalOf[F,impl.UpStreamState](impl.UpStreamState(done = false, err = None))).flatMap { upState =>
+          Stream.eval(async.signalOf[F,impl.DownStreamState](impl.Ready(None))).flatMap { dnState =>
+            impl.processInput(source,queue,upState,dnState).map { _ =>
+              new InputStream {
+                override def close(): Unit = impl.closeIs(upState,dnState)
+                override def read(b: Array[Byte], off: Int, len: Int): Int = impl.readIs(b,off,len,queue,dnState)
+                def read(): Int = impl.readIs1(queue,dnState)
+              }
+            }.onFinalize(impl.close(upState,dnState))
+          }}}
+  }
+
+
+
+  object impl {
+    /** state of the upstream, we only indicate whether upstream is done and if it failed **/
+    case class UpStreamState(done:Boolean,err:Option[Throwable])
+
+    case class Done(rslt:Option[Throwable]) extends DownStreamState
+    case class Ready(rem:Option[Bytes]) extends DownStreamState
+    sealed trait DownStreamState {  self =>
+      def isDone:Boolean = self match {
+        case Done(_) => true
+        case _ => false
+      }
+    }
+
+
+    /**
+      * Takes source and runs it through queue, interrupting when dnState signals stream is done.
+      * Note when the exception in stream is encountered the exception is emitted on the left to the queue
+      * and that would be the last message enqueued.
+      *
+      * Emits only once, but runs in background until either source is exhausted or `interruptWhenTrue` yields to true
+      */
+    def processInput[F[_]](
+      source:Stream[F,Byte]
+      , queue:mutable.Queue[F,Either[Option[Throwable],Bytes]]
+      , upState: mutable.Signal[F,UpStreamState]
+      , dnState: mutable.Signal[F,DownStreamState]
+    )(implicit F: Async[F]):Stream[F,Unit] = Stream.eval(F.delay {
+      def markUpstreamDone(result:Option[Throwable]):Unit = {
+        F.unsafeRunAsync(
+          F.flatMap(upState.set(UpStreamState(done = true, err = result))) { _ =>
+            queue.enqueue1(Left(result))
+          }
+        )(_ => ())
+      }
+
+
+      F.unsafeRunAsync(
```

Comment:
It's probably a little better to use `F.start` here instead of `unsafeRunAsync`.